llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -3496,21 +3496,30 @@ static SDValue matchSplatAsGather(SDValue SplatVal, MVT VT, const SDLoc &DL,
   if (SplatVal.getOpcode() != ISD::EXTRACT_VECTOR_ELT)
     return SDValue();
   SDValue Vec = SplatVal.getOperand(0);
-  // Only perform this optimization on vectors of the same size for simplicity.
   // Don't perform this optimization for i1 vectors.
   // FIXME: Support i1 vectors, maybe by promoting to i8?
-  if (Vec.getValueType() != VT || VT.getVectorElementType() == MVT::i1)
+  if (VT.getVectorElementType() == MVT::i1)
+    return SDValue();
+  // Additionally the element types should match
+  if (Vec.getSimpleValueType().getVectorElementType() !=
+      VT.getVectorElementType())
     return SDValue();
   SDValue Idx = SplatVal.getOperand(1);
   // The index must be a legal type.
   if (Idx.getValueType() != Subtarget.getXLenVT())
     return SDValue();
 
+  // FIXME: Can we use the indexes that are in-bound here instead?
```

Comment:
Nit, can you add a comment explaining that we need to check the sizes to ensure that index lies within VT?

Also can also leave this a TODO instead. FIXME makes me think of miscompiles/incorrectness

Reply:
done